### ISSUE TITLE
Resteasy reactive blocking and ContainerRequestFilter with cacheIllegalStateException

### ISSUE THREAD

**Issue by fabrii** [quarkusio/quarkus#32700]
### Describe the bug

I am currently in the process of migrating quarkus-resteasy-jackson libraries to quarkus-resteasy-reactive-jackson. Currently, our APIS are `@Blocking`, and to enforce that I am using the `@Blocking` annotation at RestApplication level.

We have a ContainerRequestFilter that internally consumes a method annotated with `@CacheResult`. We are receiving the following error:

The current thread cannot be blocked: vert.x-eventloop-thread-1: java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-1
        at io.smallrye.mutiny.operators.uni.UniBlockingAwait.await(UniBlockingAwait.java:30)

The error disappears if I remove the `@PreMatching` annotation.  I am using it as it is documented here: https://quarkus.io/guides/security-customization#jaxrs-security-context

Thanks!

### How to Reproduce?

Reproducer in https://github.com/fabrii/quarkus-playground/tree/reactive_thread_blocked (ignore the README).

### Output of `uname -a` or `ver`

MINGW64_NT-10.0-19045

### Output of `java -version`

11.0.13

### Quarkus version or git rev

2.16.5.Final

**Comment by geoand** [comment]
What happens here is that `@PreMatching` filters are always executed on the event loop, even though `@Blocking` is used on `Application`.

@stuartwdouglas @FroMage @cescoffier I think it makes sense to run all RESTEasy Reactive handlers on a worker thread (by default) when  `@Blocking` is used on `Application`, do you agree?

**Comment by cescoffier** [comment]
That's a good question. 
Using @Blocking on the application means that the user code is blocking. However it should not have an impact on 'framework' code. So, now the question is whether these hooks are considered as user or framework code. 

Note that @Blocking on the hook class should work.

**Comment by geoand** [comment]
> So, now the question is whether these hooks are considered as user or framework code

If you ask me, it's user code.

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/pull/32722 is what I have in mind

**Comment by fabrii** [comment]
I tried using `@Blocking`  at filter level as a workaround, but it is not working

**Comment by cescoffier** [comment]
That's clearly a bug...

**Comment by geoand** [comment]
Yes, currently prematching filters are always run on the event loop.

That's a bug, but it also creates the following question:

If a premarching filter is blocking, what do we do with the rest of the chain?

**Comment by fabrii** [comment]
I have another related question:

If I am using `@Blocking` at application level, do I have benefits on using quarkus-resteasy-reactive-jackson vs quarkus-resteasy-jackson? Is quarkus-resteasy-jackson going to be deprecated?

**Comment by jnbdz** [comment]
Still reproduces on current main. A `@PreMatching` `ContainerRequestFilter` runs on the event loop even when the class is annotated `@Blocking`:

```
blocking=vert.x-eventloop-thread-0 plain=vert.x-eventloop-thread-0 endpoint=executor-thread-1
```

(`blocking` = a `@PreMatching @Blocking` filter, `plain` = the same filter without `@Blocking`, both in one deployment.)

That matches @fabrii's report above that `@Blocking` at filter level is not a workaround.

Some findings that I think change the shape of this issue:

**1. The threading itself is documented.** `rest.adoc` says pre-matching filters "are *always* run on an event-loop thread", so this part is intentional rather than a defect.

**2. The real inconsistency is that `@Blocking` on a filter class is silently ignored.** In `ResteasyReactiveInterceptorScanner`, `@NonBlocking` on a filter class is read and sets `nonBlockingRequired`, but `@Blocking` is never read from a filter class at all — it is only consulted for endpoint methods and for the `Application` class. Meanwhile the annotation-based API rejects it at build time:

```
Wrong usage(s) of @Blocking found:
	- Filters.filter()
The @Blocking, @NonBlocking and @RunOnVirtualThread annotations may only be used on "entrypoint" methods
```

So of the three, one is honoured, one is a build error, and one is a silent no-op.

**3. There is no `blocking` counterpart in the model.** `ResourceInterceptor` only carries `nonBlockingRequired`, and `RuntimeDeploymentManager` passes just that plus `withFormRead` into `ResourceRequestFilterHandler`, which invokes the filter inline. There is currently nowhere to express "this filter needs a worker thread".

**4.** #32722 is closed and labelled `triage/invalid`, so I assume the "run every handler on a worker thread when `@Blocking` is on the `Application`" direction is off the table.

@geoand you asked in this thread: *if a pre-matching filter is blocking, what do we do with the rest of the chain?* Options as I see them:

- **A. Fail the build** when `@Blocking` is on a filter class, matching what the annotation-based API already does. Smallest change, removes the silent no-op, but leaves users of the security-customization guide with no option.
- **B. Honour it:** add a blocking flag to `ResourceInterceptor` and, once a pre-matching filter requires it, run that filter and the remainder of the pre-match chain on a worker thread. This answers the chain question by switching once and staying switched, at the cost of making pre-match threading conditional.
- **C. Docs only:** state in `security-customization.adoc` that a `@PreMatching` filter cannot block, since that guide is what leads people here.

I have the reproducer and can do A, B or C, but I would rather get direction than guess, given #32722. Happy to do C regardless, since the security guide currently sends users straight into this.
